### ISSUE TITLE
How to fix dynamic axes when exporting to ONNX?

### ISSUE THREAD

**Issue by LuHaorui-ayanami** [Maelic/SGG-Benchmark#93]
I am trying to export the model to ONNX format. However, I have a requirement where dynamic axes are not supported in my deployment environment. I need to fix the dynamic dimensions (specifically batch_size, num_objects, and num_rels) to static values during the export process.

Currently, the export function defines these dimensions as dynamic. Could you please guide me on how to modify the code to export a fully static ONNX model?

def export(model, sample_input, onnx_path, is_yolo=False, opset=17):
    # The webcam demo always letterboxes to the model's fixed input size (e.g. 640×640),
    # so spatial dimensions can be static.  Only batch size and output counts are dynamic.
    dynamic_axes = {'input': {0: 'batch_size'}}
    output_names = ['output']
    if is_yolo:
        output_names = ['boxes', 'rels']
        dynamic_axes['boxes'] = {0: 'num_objects'}
        dynamic_axes['rels'] = {0: 'num_rels'}

Any advice would be appreciated!

**Comment by LuHaorui-ayanami** [comment]
Hi maintainers, I've been exploring the source code and have some questions regarding the end-to-end (e2e) mode.

While I see the code supports an end2end flag to bypass the model's internal NMS operation, I'm unsure about its current functional status and specific use cases.

Is the end-to-end mode currently functional?
Can the codebase be run in a true end-to-end mode (no NMS), and are there any usage examples or documentation available?

Is there a YOLOv26 pre-trained model on the PSG dataset?
If the end-to-end mode is viable, do you have a pre-trained model for yolo26 (or similar) trained specifically on the Panoptic Scene Graph (PSG) dataset that you could share?

Thanks for your time and assistance!

**Comment by Maelic** [comment]
> I am trying to export the model to ONNX format. However, I have a requirement where dynamic axes are not supported in my deployment environment. I need to fix the dynamic dimensions (specifically batch_size, num_objects, and num_rels) to static values during the export process.
> 
> Currently, the export function defines these dimensions as dynamic. Could you please guide me on how to modify the code to export a fully static ONNX model?
> 
> def export(model, sample_input, onnx_path, is_yolo=False, opset=17): # The webcam demo always letterboxes to the model's fixed input size (e.g. 640×640), # so spatial dimensions can be static. Only batch size and output counts are dynamic. dynamic_axes = {'input': {0: 'batch_size'}} output_names = ['output'] if is_yolo: output_names = ['boxes', 'rels'] dynamic_axes['boxes'] = {0: 'num_objects'} dynamic_axes['rels'] = {0: 'num_rels'}
> 
> Any advice would be appreciated!

Modifying `batch_size` is trivial — just drop it from `dynamic_axes` since your `sample_input` already fixes it during tracing, no code changes needed.

`num_objects`/`num_rels` are different though — they come from real boolean-mask filtering deep in the model (`mask = preds[..., 4] > self.conf_thres` in `yolo.py::postprocess`, and `obj_mask`/`valid_rel_mask` indexing in `generate_detect_sg`), so the row count is a genuine runtime value that varies per image. Just removing them from `dynamic_axes` won't make the model static — `torch.jit.trace` will silently bake in whatever count happened to occur for your one sample image, giving you a broken model on any other input.

To get truly static shapes, you'd need to change those two spots to always emit a fixed number of slots (e.g. `max_det`) and zero out/flag invalid entries via the score column instead of dropping rows — the same "always keep `max_det` slots" pattern the codebase already uses in `get_topk_index`. Then you filter by score on the client side after inference instead of inside the graph (similarly to how the ultralytics YOLO codebase handles such cases). Once that's in place, `num_objects`/`num_rels` become true constants and you can safely drop them from `dynamic_axes`.

**Comment by Maelic** [comment]
> Hi maintainers, I've been exploring the source code and have some questions regarding the end-to-end (e2e) mode.
> 
> While I see the code supports an end2end flag to bypass the model's internal NMS operation, I'm unsure about its current functional status and specific use cases.
> 
> Is the end-to-end mode currently functional? Can the codebase be run in a true end-to-end mode (no NMS), and are there any usage examples or documentation available?
> 
> Is there a YOLOv26 pre-trained model on the PSG dataset? If the end-to-end mode is viable, do you have a pre-trained model for yolo26 (or similar) trained specifically on the Panoptic Scene Graph (PSG) dataset that you could share?
> 
> Thanks for your time and assistance!

End-to-end models are fully supported in the codebase, such as YOLOV10, YOLO26 etc... To use an end-to-end model, you don't have to change anything, the NMS will be bypass automatically given the type of architecture and the fallback will be: https://github.com/Maelic/SGG-Benchmark/blob/a6642012e66047c3544bed8784bbf9acdc87411b/sgg_benchmark/modeling/backbone/yolo.py#L141-L149
where the indices are automatically extracted from the final array, so it works as well with DAMP feature extraction like in REACT++.

I have uploaded a YOLO26m checkpoint trained on PSG for you there: https://drive.google.com/file/d/1UOAlPtNczwTDH8ZFPitVc1ujZ2Gnpv4w/view?usp=sharing

You can thank me by citing my papers in your work :)

Happy coding!
